### PR TITLE
feat: add gen

### PR DESCRIPTION
**Pull request by sashamelentyev**

### CODE REVIEW

**Comment by sashamelentyev** [comment]
@ldez hello!
I added generation, please review this. 
Thank you!

**Comment by sashamelentyev** [comment]
Changed map to array in gen data, because map have random iteration order

**Comment by ldez** [comment]
The implementation seems not right: using a hard-coded mapping to create a hard-coded mapping is not really useful.

I think instead of trying to generate the mapping and the tests inside the same pass, you have to create 2 passes/"main"/packages: one for tests (that use the real mapping) and one for mapping (this one is optional if you are not able to not use mapping to generate mapping)

Tips: to be able to use the mapping inside the generator and the analyzer without publically exporting the global vars you can just move the mapping file to the internal package and export the global vars.

**Comment by sashamelentyev** [comment]
@ldez thank you for review!
I hard-code mapping, and move file to `internal/mapping`. And move `internal` to `pkg/analyzer`.
Right now generate tests by real mapping only

**Comment by ldez** [comment]
another tip:

```go
// gofmt
source, err := format.Source(buffer.Bytes())
if err != nil {
	return fmt.Errorf("source: %w", err)
}
```

**Comment by sashamelentyev** [comment]
@Idez, wow, thank you!
You beast!
Add this. 
Can I merge this?

File: pkg/analyzer/internal/gen.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+	execute("pkg/analyzer/testdata/src/a/time/month.go", t, "template.tmpl", tmpl{PackageName: "time_test", Mapping: mapping.TimeMonth})
+	execute("pkg/analyzer/testdata/src/a/time/layout.go", t, "template.tmpl", tmpl{PackageName: "time_test", Mapping: mapping.TimeLayout})
+}
+
+func execute(fileName string, t *template.Template, templateName string, data any) {
+	f, err := os.Create(fileName)
+	if err != nil {
+		log.Fatalln(err)
+	}
+	defer func() { _ = f.Close() }()
```
Comment (ldez) [inline pkg/analyzer/internal/gen.go:46]:
```suggestion
```

**Comment by ldez** [comment]
Ldez not idez :wink:

**Comment by sashamelentyev** [comment]
My bad

**Comment by ldez** [comment]
I have several suggestions.

I think you can rename template files following this pattern `test-httpmethod.go.tmpl`.

The code generator can be written differently:

<details>

```go
func main() {
	t := template.Must(
		template.New("template").
			Funcs(map[string]any{"quoteMeta": regexp.QuoteMeta}).
			ParseFS(templateDir, "template/*.tmpl"),
	)

	operations := []struct {
		templateName string
		packageName  string
		mapping      map[string]string
		fileName     string
	}{
		{
			templateName: "template.tmpl",
			packageName:  "crypto_test",
			mapping:      mapping.CryptoHash,
			fileName:     "pkg/analyzer/testdata/src/a/crypto/crypto.go",
		},
		{
			templateName: "httpmethod.tmpl",
			packageName:  "http_test",
			mapping:      mapping.HTTPMethod,
			fileName:     "pkg/analyzer/testdata/src/a/http/method.go",
		},
		{
			templateName: "httpstatuscode.tmpl",
			packageName:  "http_test",
			mapping:      mapping.HTTPStatusCode,
			fileName:     "pkg/analyzer/testdata/src/a/http/statuscode.go",
		},
		{
			templateName: "template.tmpl",
			packageName:  "rpc_test",
			mapping:      mapping.DefaultRPCPath,
			fileName:     "pkg/analyzer/testdata/src/a/rpc/rpc.go",
		},
		{
			templateName: "template.tmpl",
			packageName:  "time_test",
			mapping:      mapping.TimeWeekday,
			fileName:     "pkg/analyzer/testdata/src/a/time/weekday.go",
		},
		{
			templateName: "template.tmpl",
			packageName:  "time_test",
			mapping:      mapping.TimeMonth,
			fileName:     "pkg/analyzer/testdata/src/a/time/month.go",
		},
		{
			templateName: "template.tmpl",
			packageName:  "time_test",
			mapping:      mapping.TimeLayout,
			fileName:     "pkg/analyzer/testdata/src/a/time/layout.go",
		},
	}

	for _, operation := range operations {
		data := map[string]any{
			"PackageName": operation.packageName,
			"Mapping":     operation.mapping,
		}

		err := execute(t, operation.templateName, data, operation.fileName)
		if err != nil {
			log.Fatal(err)
		}
	}
}

func execute(t *template.Template, templateName string, data any, fileName string) error {
	var builder bytes.Buffer
	if err := t.ExecuteTemplate(&builder, templateName, data); err != nil {
		return err
	}

	sourceData, err := format.Source(builder.Bytes())
	if err != nil {
		return err
	}

	return os.WriteFile(fileName, sourceData, os.ModePerm)
}

```

</details>

It's a bit more verbose but a lot more readable from my point of view.

**Comment by ldez** [comment]
Note: you always say "forgive me", but I think this is not what you what to say.

"forgive me" means "can you forget me", but I don't want to forget you :smile: 

Maybe you are trying to say: "excuse me" or "sorry"

**Comment by sashamelentyev** [comment]
Thank you! I take your suggestions. 

Oh, yes, I mean "excuse me". Thank you for English lesson!

"Thank you for all" x 1000

File: pkg/analyzer/internal/gen.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+		template.New("template").
+			Funcs(map[string]any{"quoteMeta": regexp.QuoteMeta}).
+			ParseFS(templateDir, "template/*.tmpl"),
+	)
+
+	operations := []struct {
+		templateName string
+		packageName  string
+		mapping      map[string]string
+		fileName     string
```
Comment (ldez) [inline pkg/analyzer/internal/gen.go:37]:
```suggestion
		mapping      map[string]string
		packageName  string
		templateName string
		fileName     string
```

File: pkg/analyzer/internal/gen.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+	}{
+		{
+			templateName: "test-template.go.tmpl",
+			packageName:  "crypto_test",
+			mapping:      mapping.CryptoHash,
+			fileName:     "pkg/analyzer/testdata/src/a/crypto/crypto.go",
```
Comment (ldez) [inline pkg/analyzer/internal/gen.go:43]:
```suggestion
			mapping:      mapping.CryptoHash,
			packageName:  "crypto_test",
			templateName: "test-template.go.tmpl",
			fileName:     "pkg/analyzer/testdata/src/a/crypto/crypto.go",
```

File: pkg/analyzer/internal/gen.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+		},
+		{
+			templateName: "test-httpmethod.go.tmpl",
+			packageName:  "http_test",
+			mapping:      mapping.HTTPMethod,
+			fileName:     "pkg/analyzer/testdata/src/a/http/method.go",
```
Comment (ldez) [inline pkg/analyzer/internal/gen.go:49]:
```suggestion
			mapping:      mapping.HTTPMethod,
			packageName:  "http_test",
			templateName: "test-httpmethod.go.tmpl",
			fileName:     "pkg/analyzer/testdata/src/a/http/method.go",
```

File: pkg/analyzer/internal/gen.go [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+		},
+		{
+			templateName: "test-httpstatuscode.go.tmpl",
+			packageName:  "http_test",
+			mapping:      mapping.HTTPStatusCode,
+			fileName:     "pkg/analyzer/testdata/src/a/http/statuscode.go",
```
Comment (ldez) [inline pkg/analyzer/internal/gen.go:55]:
```suggestion
			mapping:      mapping.HTTPStatusCode,
			packageName:  "http_test",
			templateName: "test-httpstatuscode.go.tmpl",
			fileName:     "pkg/analyzer/testdata/src/a/http/statuscode.go",
```

File: pkg/analyzer/internal/gen.go [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+		},
+		{
+			templateName: "test-template.go.tmpl",
+			packageName:  "rpc_test",
+			mapping:      mapping.DefaultRPCPath,
+			fileName:     "pkg/analyzer/testdata/src/a/rpc/rpc.go",
```
Comment (ldez) [inline pkg/analyzer/internal/gen.go:61]:
```suggestion
			mapping:      mapping.DefaultRPCPath,
			packageName:  "rpc_test",
			templateName: "test-template.go.tmpl",
			fileName:     "pkg/analyzer/testdata/src/a/rpc/rpc.go",
```

File: pkg/analyzer/internal/gen.go [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+		},
+		{
+			templateName: "test-template.go.tmpl",
+			packageName:  "time_test",
+			mapping:      mapping.TimeWeekday,
+			fileName:     "pkg/analyzer/testdata/src/a/time/weekday.go",
```
Comment (ldez) [inline pkg/analyzer/internal/gen.go:67]:
```suggestion
			mapping:      mapping.TimeWeekday,
			packageName:  "time_test",
			templateName: "test-template.go.tmpl",
			fileName:     "pkg/analyzer/testdata/src/a/time/weekday.go",
```

File: pkg/analyzer/internal/gen.go [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+		},
+		{
+			templateName: "test-template.go.tmpl",
+			packageName:  "time_test",
+			mapping:      mapping.TimeMonth,
+			fileName:     "pkg/analyzer/testdata/src/a/time/month.go",
```
Comment (ldez) [inline pkg/analyzer/internal/gen.go:73]:
```suggestion
			mapping:      mapping.TimeMonth,
			packageName:  "time_test",
			templateName: "test-template.go.tmpl",
			fileName:     "pkg/analyzer/testdata/src/a/time/month.go",
```

File: pkg/analyzer/internal/gen.go [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+		},
+		{
+			templateName: "test-template.go.tmpl",
+			packageName:  "time_test",
+			mapping:      mapping.TimeLayout,
+			fileName:     "pkg/analyzer/testdata/src/a/time/layout.go",
```
Comment (ldez) [inline pkg/analyzer/internal/gen.go:79]:
```suggestion
			mapping:      mapping.TimeLayout,
			packageName:  "time_test",
			templateName: "test-template.go.tmpl",
			fileName:     "pkg/analyzer/testdata/src/a/time/layout.go",
```

File: pkg/analyzer/internal/gen.go [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+	"regexp"
+	"text/template"
+
+	"github.com/sashamelentyev/usestdlibvars/pkg/analyzer/internal/mapping"
+)
+
+type tmpl struct {
+	PackageName string
+	Mapping     map[string]string
+}
```
Comment (ldez) [inline pkg/analyzer/internal/gen.go:21]:
```suggestion
```

**Comment by ldez** [comment]
you can pick all the changes in one commit in the "Files changed" tab

**Comment by ldez** [comment]
A more general tip: you can choose the kind of merge you want to use on the "merge button".

You can use the "Squash merge" instead of "Create a merge commit"

**Comment by sashamelentyev** [comment]
Thank you for tips!

**Comment by ldez** [comment]
LGTM

**Pull request merged**